File: lilac/src/main/java/com/hartwig/hmftools/lilac/variant/LilacVCF.java
package com.hartwig.hmftools.lilac.variant;

import com.hartwig.hmftools.lilac.hla.HlaAllele;

import htsjdk.variant.variantcontext.VariantContext;
import htsjdk.variant.variantcontext.VariantContextBuilder;
import htsjdk.variant.variantcontext.writer.VariantContextWriter;
import htsjdk.variant.variantcontext.writer.VariantContextWriterBuilder;
import htsjdk.variant.vcf.VCFFileReader;
import htsjdk.variant.vcf.VCFHeader;
import htsjdk.variant.vcf.VCFHeaderLine;
import htsjdk.variant.vcf.VCFHeaderLineCount;
import htsjdk.variant.vcf.VCFHeaderLineType;
import htsjdk.variant.vcf.VCFInfoHeaderLine;

import java.io.File;
import java.util.List;
import java.util.Set;
import java.util.StringJoiner;

public class LilacVCF
{
    private final VCFFileReader mHeader;
    private final VariantContextWriter mWriter;

    public static final String HLA = "HLA";

    public LilacVCF(final String outputVCF, final String templateVCF)
    {
        mHeader = new VCFFileReader(new File(templateVCF), false);

        mWriter = new VariantContextWriterBuilder()
                .setReferenceDictionary(mHeader.getFileHeader().getSequenceDictionary())
                .setOutputFile(outputVCF)
                .build();
    }

    public final LilacVCF writeHeader(final String lilacVersion)
    {
        VCFHeader newHeader = new VCFHeader(mHeader.getFileHeader());
        newHeader.addMetaDataLine(new VCFHeaderLine("lilacVersion", lilacVersion));
        newHeader.addMetaDataLine(new VCFInfoHeaderLine(HLA, VCFHeaderLineCount.UNBOUNDED, VCFHeaderLineType.String, "HLA Type"));
        mWriter.writeHeader(newHeader);
        return this;
    }

    public final void writeVariant(final VariantContext context, final List<HlaAllele> alleles)
    {
        String alleleStr = alleles.isEmpty() ? "UNKNOWN" : HlaAllele.toString(alleles);
        VariantContext newContext = new VariantContextBuilder(context).attribute(HLA, alleleStr).make();
        mWriter.add(newContext);
    }

    public void close()
    {
        this.mWriter.close();
    }
}

File: lilac/src/main/kotlin/com/hartwig/hmftools/lilac/variant/LilacVCF.kt
package com.hartwig.hmftools.lilac.variant

import com.hartwig.hmftools.lilac.LilacApplication
import com.hartwig.hmftools.lilac.hla.HlaAllele
import htsjdk.variant.variantcontext.VariantContext
import htsjdk.variant.variantcontext.VariantContextBuilder
import htsjdk.variant.variantcontext.writer.VariantContextWriterBuilder
import htsjdk.variant.vcf.*
import java.io.File

class LilacVCF(outputVCF: String, templateVCF: String) : AutoCloseable {
    companion object {
        const val HLA = "HLA"
    }

    private val header = VCFFileReader(File(templateVCF), false).use { it.fileHeader }
    private val writer = VariantContextWriterBuilder()
            .setReferenceDictionary(header.sequenceDictionary)
            .setOutputFile(outputVCF)
            .build()

    fun writeHeader(): LilacVCF {
        val newHeader = VCFHeader(header)
        newHeader.addMetaDataLine(VCFHeaderLine("lilacVersion", LilacApplication.VERSION.version()))
        newHeader.addMetaDataLine(VCFInfoHeaderLine(HLA, VCFHeaderLineCount.UNBOUNDED, VCFHeaderLineType.String, "HLA Type"))

        writer.writeHeader(newHeader)
        return this
    }

    fun writeVariant(context: VariantContext, alleles: Set<HlaAllele>) {
        val hlaLabel = if (alleles.isEmpty()) {
            listOf("UNKNOWN")
        } else {
            alleles.map { it.toString() }
        }

        val newContext = VariantContextBuilder(context).attribute(HLA, hlaLabel).make()
        writer.add(newContext)
    }

    override fun close() {
        writer.close()
    }

}